File: top_interview_150/080_remove_duplicates.ts
/**
 * Given an integer array nums sorted in non-decreasing order,
 * remove some duplicates in-place such that each unique element appears at most twice.
 * Return k after placing the final result in the first k slots of nums.
 */

function removeDuplicates(nums: number[]): number {
  if (nums.length < 3) {
    return nums.length;
  }

  // start from the second index

  let uniqueCount = 2;

  for (let i = 2; i < nums.length; i++) {
    if (nums[i] !== nums[i - 1] && nums[i] !== nums[i - 2]) {
      console.log({ uniqueCount, nums });
      nums[uniqueCount] = nums[i];
      uniqueCount++;
    }
  }

  console.log("FINAL");
  console.log({ uniqueCount, nums });

  return uniqueCount;
}

export { removeDuplicates };

File: top_interview_150/080_remove_duplicates.rb
# @param {Integer[]} nums
# @return {Integer}
def remove_duplicates(nums)
  l = 0
  r = 0

  while r < nums.length
    count = 1

    while r + 1 < nums.length && nums[r] == nums[r + 1]
      r += 1
      count += 1
    end

    [2, count].min.times do
      nums[l] = nums[r]
      l += 1
    end

    r += 1
  end

  l
end
